### PR TITLE
New Language Server API Implementations / Mocks

### LINKED ISSUE #1776: Implement the Package Manager API to use for the IDE
### Summary

The API that is to be designed in #1764 has to be implemented and connected with the package-manager component.

### Value

- The IDE can use the package-manager.

### Specification

- [ ] The API has to be implemented as described in the documentation.
- [ ] The endpoints need to be connected to the package-manager component.
- [ ] Implement logic related to the endpoints:
  - [ ] Listing dependencies of the project (probably can be done by looking at the loaded libraries or alternatively may require parsing the files).
  - [ ] Modifying edition settings.
  - [ ] Notifications for downloads integrated with imports.
  - [ ] Other designed endpoints.
  - This does not include handling the content roots which have their own task: #1780.

### Acceptance Criteria & Test Cases

- [ ] The API is tested as much as reasonably possible.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Closes #1776 

### Important Notes

- Errors are currently handled in a simplified manner, without distinction to different causes (the message includes the cause, but the error code is currently always the same). This will be improved in #1877.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [ ] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/libraries/EditionReference.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      Json.obj(CodecField.Type -> CodecType.CurrentProjectEdition.asJson)
+  }
+
+  implicit val decoder: Decoder[EditionReference] = { json =>
+    val typeCursor = json.downField(CodecField.Type)
+    typeCursor.as[String].flatMap {
+      case CodecType.NamedEdition =>
+        for {
+          editionName <- json.get[String](CodecField.EditionName)
+        } yield NamedEdition(editionName)
```
Comment (4e6) [inline engine/language-server/src/main/scala/org/enso/languageserver/libraries/EditionReference.scala]: you can derive decoders for case classes with `io.circe.generic.auto._` and then do `Decoder[NamedEdition].tryDecode(cursor)`. Check `ContextRegistryProtocol` for examples

File: RELEASES.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Enso Next
 
+## Tooling
+
+- Implement parts of the new Language Server API related to library support
+  ([#1875](https://github.com/enso-org/enso/pull/1875)). Parts of the API are
+  still mocked, but they are supported too for testing purposes.
```
Comment (iamrecursion) [inline RELEASES.md]:
```suggestion
  still mocked internally, but they are supported externally for testing purposes.
```

File: build.sbt [hunk 3]
```diff
@@ -716,6 +716,20 @@ lazy val cli = project
     Test / parallelExecution := false
   )
 
+lazy val `task-progress-notifications` = project
```
Comment (iamrecursion) [inline build.sbt:720]: You need to add this to the global project.

File: engine/language-server/src/main/scala/org/enso/languageserver/libraries/LibraryApi.scala [hunk 4]
```diff
@@ -0,0 +1,235 @@
+package org.enso.languageserver.libraries
+
+import io.circe.Json
+import io.circe.literal.JsonStringContext
+import org.enso.editions.{LibraryName, LibraryVersion}
+import org.enso.jsonrpc.{Error, HasParams, HasResult, Method, Unused}
+
+object LibraryApi {
```
Comment (4e6) [inline engine/language-server/src/main/scala/org/enso/languageserver/libraries/LibraryApi.scala:8]: I assume the new API is documented.

**Review by 4e6 — APPROVED** [review]
Looks good

File: lib/scala/distribution-manager/src/main/scala/org/enso/distribution/EditionManager.scala [hunk 5]
```diff
@@ -38,10 +41,46 @@ class EditionManager(searchPaths: List[Path]) {
...
-  def resolveEngineVersion(
-    edition: Option[Editions.RawEdition]
-  ): Try[EnsoVersion] =
-    edition
-      .map(engineVersionResolver.resolveEnsoVersion(_).toTry)
-      .getOrElse(Success(DefaultEnsoVersion))
+  def resolveEngineVersion(edition: Editions.RawEdition): Try[SemVer] =
+    engineVersionResolver.resolveEnsoVersion(edition).toTry
+
+  // TODO [RW] download edition updates
```
Comment (iamrecursion) [inline lib/scala/distribution-manager/src/main/scala/org/enso/distribution/EditionManager.scala]: Link this to an issue if you can.

File: lib/scala/pkg/src/main/scala/org/enso/pkg/Config.scala [hunk 6]
```diff
@@ -205,7 +205,26 @@ object Config {
...
+      *
+      * `ensoVersion` is dropped, because due to migration it is overridden by
+      * `edition` and we don't want to have both to avoid inconsistency.
+      *
+      * `prefer-local-libraries` cannot be inherited, because if it was set to
+      * `true` and we have changed it to `false`, overrides will not include it,
+      * because, as `false` is its default value, we just ignore the field. But
+      * if we inherit it from original fields, we would get `true` back. If the
+      * setting is still set to true, it will be included in the overrides, so
+      * it does not have to be inherited either.
```
Comment (iamrecursion) [inline lib/scala/pkg/src/main/scala/org/enso/pkg/Config.scala:219]: Great explanation for something potentially confusing.

**Review by iamrecursion — APPROVED** [review]
Approved, conditional on @4e6 doing a full review.

File: engine/language-server/src/main/scala/org/enso/languageserver/libraries/LibraryApi.scala (hunk 4 above)
Comment (radeusgd) [reply engine/language-server/src/main/scala/org/enso/languageserver/libraries/LibraryApi.scala:8]: Yes, it is documented in `protocol-language-server.md#library-related-operations`.

File: build.sbt (hunk 3 above)
Comment (radeusgd) [reply build.sbt:720]: Pheew, thanks for catching that, I don't know why I'm always forgetting this...

**Pull request merged**